Displaying file upload strain errors all at once

vlagrassa opened the pull request:
- Validators are now split into two functions: the `step` and the `final`
  - For `validate_strains`, the `step` keeps a list of invalid lines, categorized by error type: wrong species, unknown strain, blank strain, and (optionally) duplicate strain
  - For `validate_strains`, the `final` runs through each category, and displays the first one that has one or more errors
  - For the other validators, only the `step` is required (remains unchanged)
- Error messages are meant to be human-readable
  - If 1-3 strains are invalid, display those in an inline list (using correct pluralization, commas + "and", etc)
  - If 4+ are invalid, include a dropdown (a BS `collapse` element) with the full list
  - Species names are italicized
- Submission API calls accept an optional URL variable `reloadonerr`
  - If `0`, returns the error message without using Flask `flash`. The JS function will manually show the error without reloading the page
  - If `1` or unset, uses `flash` to display the message. The JS function will reload the page to make this appear

vlagrassa committed:
Show file upload validation errors without reloading page

vlagrassa committed:
Display all invalid headers at once

vlagrassa committed:
Report all invalid strains at once (first draft)

Keeps a running list of invalid strains in three categories:
- empty string
- wrong species
- unknown strain

Then goes through those categories in order, outputting all results at once.  Truncates to first 8 results (this can/should change).

Should this also track duplicate rows if applicable?

vlagrassa committed:
Split validators into "step" and "final" functions

vlagrassa committed:
If more than 4 strain names invalid, list in dropdown in error msg

vlagrassa committed:
Check for wrong species first

vlagrassa committed:
Abstract out check_strain_list logic

vlagrassa committed:
Rename "many" to "default" and use as fallback (+docs)

vlagrassa committed:
Handle duplicate strains in final validator step too

vlagrassa committed:
Italicize species names in alert messages

vlagrassa committed:
Merge branch 'development' into enhancement/displaying-errors

r-vieira commented on src/modules/site-v2/base/utils/tools.py:
```diff
@@ -122,7 +122,11 @@ def try_submit(EntityClass, user, data, no_cache):
+      'message': msg,
+      'full_msg_link': ex.full_msg_link,
+      'full_msg_body': ex.full_msg_body,
+    }, 400
```
`400` sounds like a good candidate for an `enum/class`. I.e.: `HttpStatusCode.BAD_REQUEST`

vlagrassa replied:
Happy to make the enum if you think it would help...  I'm tempted to say the code numbers are pretty transparent, but I can see how they might be confusing, esp. when they're returned a few function calls down (as opposed to from the view function directly)

r-vieira replied:
I see -- maybe I'm looking at this from the perspective of new developer approaching the codebase where it might be useful to know the domain of the 'magic numbers' in code, eg: is the number referring to some internal status code of the application layer?Http? others? in this case, it helps to clarify that we are setting the HttpStatusCode semantically in the use of enums (of course, at the expense of extra verbosity 😬).

vlagrassa replied:
Makes sense. I'll add a card for it in BC

r-vieira commented on src/modules/site-v2/base/views/tools/genetic_mapping.py:
```diff
@@ -115,10 +115,12 @@ def submit():
-      # If there is an error or a ready message, flash it
-      if not code == 200:
+      # If there was an error, flash it
+      if code != 200 and int(request.args.get('reloadonerr', 1)):
```
consider `reloadonerror` instead of shorter version.

r-vieira replied:
`200` => HttpStatusCode.OK enum/class?

vlagrassa replied:
Fixed the first part in [b137d57](https://github.com/AndersenLab/CAENDR/pull/402/commits/b137d5709aff41d74bc8177241ca161ae2e3d02d)

r-vieira commented on src/modules/site-v2/templates/_scripts/submit-job.js:
```diff
@@ -18,20 +18,32 @@ $.ajaxSetup({
 #}
 {% macro def_submit_job(tool_name, as_form_data=false, func_name='submit_job') %}
-function {{func_name}}(data, modal_id, new_tab=false) {
+function {{func_name}}(data, modal_id, new_tab=false, propagate_err=true) {
```
prefer `propagate_error` for clarity over abbreviations.

r-vieira replied:
I appreciate the `c-macro-ish-ness` nature of `{{ func_name }}`. Is there a chance it might be null/undefined/use reserved keyword, etc?

vlagrassa replied:
Fixed the first part in [b137d57](https://github.com/AndersenLab/CAENDR/pull/402/commits/b137d5709aff41d74bc8177241ca161ae2e3d02d)

vlagrassa replied:
To the second point - as it turns out, none of the calling sites use the `func_name` argument to change the name.  I think I put this in so we'd have a way to manually avoid name collisions (rather than the parameter being filled in by a variable), but that ended up not being necessary...

(To clarify -- I was envisioning something like, the following, using a string:
`{{ def_submit_job( ..., func_name='submit_job_foobar' ) }}`

...rather than something using a variable:
`{{ def_submit_job( ..., func_name=foobar ) }}`

...but since we don't use either, I could also just remove the parameter)

r-vieira replied:
got it, thanks!

r-vieira commented on src/modules/site-v2/templates/_scripts/utils.js:
```diff
@@ -98,3 +98,54 @@ function force_download(e) {
+    node.firstElementChild.innerText += ' ';
+
+    // Create a unique ID to link the trigger & collapse elements
+    const id = `err-dropdown-${(new Date()).getTime()}`;
```
prefer `error` vs `err`

vlagrassa replied:
Fixed in [b137d57](https://github.com/AndersenLab/CAENDR/pull/402/commits/b137d5709aff41d74bc8177241ca161ae2e3d02d)

r-vieira commented on src/modules/site-v2/templates/tools/pairwise_indel_finder/submit.html:
```diff
@@ -549,7 +549,7 @@ <h1 class="modal-title fs-5" id="generate-primers-modal-label">Thank You</h1>
 
   // Submit the job, opening results in a new tab if applicable
-  submit_job(data, '#generate-primers-modal', true)
+  submit_job(data, '#generate-primers-modal', true, false)
```
consider using a config object instead of sequencial booleans, (...,..., { foo: true, bar: false}) for clarity of order.

vlagrassa replied:
Fixed in [27fe543](https://github.com/AndersenLab/CAENDR/pull/402/commits/27fe543f3e581ff847e79c2b5470b48b31e2d903)

r-vieira reviewed (changes requested):
review inline comments and lmk what you think.

vlagrassa committed:
Expand name abbreviations for clarity

- `reloadonerr` -> `reloadonerror`
- `propagate_err` -> `propagate_error`
- `err-dropdown` -> `error-dropdown`

vlagrassa committed:
Use config object argument for clarity
